store/redis-store.ts

```diff
@@ -25,72 +45,93 @@ export class RedisStore<Value> extends BaseStore<Value> {
    * @param expirySeconds -1 for persistent cache, 0 for no cache, > 0 for expiry in seconds
    * @returns value if present in cache
    */
-  async getOrSet<V = Value>(
+  async getOrSet(
     key: string,
-    valueFn: () => V | Promise<V>,
+    valueFn: () => Value | Promise<Value>,
     expirySeconds: number
-  ): Promise<V> {
+  ): Promise<Value> {
     this.startCacheUpdater({ key, valueFn, expirySeconds })
 
-    // do not read cache if expiry is 0
-    const read = expirySeconds !== 0 ? await this.get<V>(key) : undefined
+    const fetchAndStore = async () => {
+      const result = valueFn()
 
-    let valuePromise = this._promises.get(key)
-    if (read !== undefined) {
-      debug('RedisStore.getOrSet: returning the value present in storage')
-      return read
-    } else if (valuePromise) {
-      debug('RedisStore.getOrSet: value being queried already, waiting for it')
-      return valuePromise
-    } else {
-      debug('RedisStore.getOrSet: value not present in storage, fetching it')
-      valuePromise = valueFn()
-      if (valuePromise instanceof Promise) {
-        this._promises.set(key, valuePromise)
+      if (result instanceof Promise) {
+        this._promises.set(key, result)
       }
+
       try {
-        const value = await valuePromise
+        const value = await result
         // override expirySeconds if cacheSeconds is provided
         if (typeof value.cacheSeconds === 'number') {
           expirySeconds = value.cacheSeconds
         }
         // do not write to cache if expiry is 0
         if (expirySeconds !== 0) {
-          await this.set<V>(key, value, expirySeconds)
+          await this.set(key, value, expirySeconds)
         }
-        this._promises.set(key, undefined)
+
+        this._promises.delete(key)
+
         return value
-      } catch (e) {
-        this._promises.set(key, undefined)
-        throw e
+      } catch (err) {
+        this._promises.delete(key)
+
+        throw err
       }
     }
+    // do not read cache if expiry is 0
+    const cachedValue = expirySeconds !== 0 ? await this.get(key) : undefined
+
+    if (cachedValue) {
+      debug('RedisStore.getOrSet: returning the value present in storage')
+
+      // return from cache and initiate cache update
+      const isExpired =
+        currentTimestamp() - cachedValue.cacheTimestamp >= cachedValue.cacheSeconds
+      if (isExpired) fetchAndStore()
```

Comment:
There might be an uncaught error here, either we spit it into std err output by console.error or over email or discord.

Also maybe we should also store the error if existing cache is old enough. Imagine stale prices being served, vs api errors with 500. The incident might go unnoticed until it's very late, but if API errors out at least we get bug reports on discord by users.